### ISSUE TITLE
Graphene Module won't load (OS X El Capitan, Node 4.1.1)

### ISSUE THREAD

**Issue by arik-so** [PeculiarVentures/graphene#41]
I am trying to run the following code:

```javascript
var graphene = require('graphene-pk11');
var Module = graphene.Module;

var mod = Module.load('/usr/local/lib/softhsm/libsofthsm2.so', 'SoftHSM');
// var mod = Module.load('/usr/local/Cellar/softhsm/2.0.0/lib/softhsm/libsofthsm2.so', 'SoftHSM');
mod.initialize();
```

However, in line 4, it crashes with the following error:

> /Users/project-folder/node_modules/graphene-pk11/node_modules/ffi/lib/dynamic_library.js:74
>     throw new Error('Dynamic Linking Error: ' + err)
>     ^
> 
> Error: Dynamic Linking Error: dlopen(/usr/local/lib/softhsm/libsofthsm2.so.dylib, 2): image not found
>     at new DynamicLibrary (/Users/project-folder/node_modules/graphene-pk11/node_modules/ffi/lib/dynamic_library.js:74:11)
>     at Object.Library (/Users/project-folder/node_modules/graphene-pk11/node_modules/ffi/lib/library.js:45:12)
>     at new Pkcs11 (/Users/project-folder/node_modules/graphene-pk11/build/pkcs11/pkcs11.js:14:24)
>     at Function.Module.load (/Users/project-folder/node_modules/graphene-pk11/build/module.js:70:19)
>     at Object.<anonymous> (/Users/project-folder/lab/hsm.js:8:18)
>     at Module._compile (module.js:434:26)
>     at Object.Module._extensions..js (module.js:452:10)
>     at Module.load (module.js:355:32)
>     at Function.Module._load (module.js:310:12)
>     at Module.runMain [as _onTimeout] (module.js:475:10)

Quite curiously, Node is complaining about not being able to find `libsofthsm2.so.dylib` even though in the method call it explicitly says `libsofthsm2.so`. What might be causing it?

Additionally, I should note that `/usr/local/lib/softhsm/libsofthsm2.so` is a symlink to `/usr/local/Cellar/softhsm/2.0.0/lib/softhsm/libsofthsm2.so`, which I have also tried using in the commented out line, with the same results.

**Comment by rmhrisk** [comment]
Looks like your on a Mac.

What files exist in the target folder?

This may also be interesting: http://stackoverflow.com/questions/2339679/what-are-the-differences-between-so-and-dylib-on-osx

**Comment by arik-so** [comment]
Yep, I'm on a Mac, El Capitan. These are the files I have in the folder, with their respective permissions:
```
-rw-r--r--  1 root  admin  14509560 Aug 12 16:32 libsofthsm2.a
-rwxr-xr-x  1 root  admin       967 Aug 12 16:32 libsofthsm2.la
-rwxr-xr-x  1 root  admin   1113188 Aug 12 16:32 libsofthsm2.so
```

**Comment by rmhrisk** [comment]
How did you install softhsm?

**Comment by arik-so** [comment]
I first tried installing it directly according to the repository readme, but that didn't work. I then installed it using homebrew, but that couldn't initialize tokens, throwing the couldn't initialize library error. Finally, I installed it using this tutorial: https://fossies.org/dox/softhsm-2.1.0/md_OSX-NOTES.html, and everything worked perfectly.

**Comment by rmhrisk** [comment]
arik-so

I just spent the day messing with Travis and have builds of SoftHSM working on OSX with Graphene.

I think the issue may be in your build of softhsm, I used these steps : https://github.com/opendnssec/SoftHSMv2/blob/develop/OSX-NOTES.md

**Comment by rmhrisk** [comment]
I am happy to help but since this isn't with Graphene I will close this issue, feel free to use it to communicate with us though.

**Comment by arik-so** [comment]
Hey Ryan, thanks for your response. I have wiped SoftHSM from my computer and reinstalled it using the instructions you linked to verbatim. SoftHSM still works perfectly, I can initialize tokens and everything, but I'm still getting the same error with Graphene, and I still don't have a `libsofthsm.dylib` or `libsofthsm.so.dylib` file, though SoftHSM appears to be doing just fine without it. I only have `libsofthsm.so`. The file structure appears to still be the same as it was when I installed SoftHSM using brew.

**Comment by rmhrisk** [comment]
This is very strange.

In my case I am using :

**Comment by arik-so** [comment]
Ok, I think we are on to something. Your method is the one that requires installing Graphene globally. When I tried that, it errored out. Previously, I have only been trying to install it in a package, where npm was not complaining. Here is the error I got trying to install Graphene globally:

> -> running install
> 
> \>\> written zero files
> typings/node/node.d.ts(87,23): error TS1110: Type expected.
> typings/node/node.d.ts(874,35): error TS1110: Type expected.
> typings/node/node.d.ts(2103,53): error TS1005: '=' expected.
> typings/node/node.d.ts(2104,77): error TS1005: '=' expected.
> typings/node/node.d.ts(2105,53): error TS1005: '=' expected.
> typings/node/node.d.ts(2105,101): error TS1005: '=' expected.
> typings/node/node.d.ts(2115,55): error TS1005: '=' expected.
> typings/node/node.d.ts(2116,75): error TS1005: '=' expected.
> typings/node/node.d.ts(2117,55): error TS1005: '=' expected.
> typings/node/node.d.ts(2117,99): error TS1005: '=' expected.
> npm ERR! Darwin 15.5.0
> npm ERR! argv "/Users/arik/.nvm/versions/node/v4.1.1/bin/node" "/Users/arik/.nvm/versions/node/v4.1.1/bin/npm" "install" "-g" "graphene-pk11"
> npm ERR! node v4.1.1
> npm ERR! npm  v2.15.6
> npm ERR! code ELIFECYCLE
> 
> npm ERR! graphene-pk11@2.0.9 install: `tsd install && tsc`
> npm ERR! Exit status 2
> npm ERR! 
> npm ERR! Failed at the graphene-pk11@2.0.9 install script 'tsd install && tsc'.
> npm ERR! This is most likely a problem with the graphene-pk11 package,
> npm ERR! not with npm itself.
> npm ERR! Tell the author that this fails on your system:
> npm ERR!     tsd install && tsc
> npm ERR! You can get information on how to open an issue for this project with:
> npm ERR!     npm bugs graphene-pk11
> npm ERR! Or if that isn't available, you can get their info via:
> npm ERR! 
> npm ERR!     npm owner ls graphene-pk11
> npm ERR! There is likely additional logging output above.
> 
> npm ERR! Please include the following file with any support request:
> npm ERR!     /Users/arik/Developer/SoftHSMv2/npm-debug.log

However, I have tsd installed globally, and both tsd and tsc are available commands.

**Comment by microshine** [comment]
Hello! Check you TypeScript version.

node:87 uses new type for BufferEncoding variable. It was supported in 1.8.0 version

**Comment by arik-so** [comment]
Thanks! Updating Typescript did the trick. My version was 1.7.x. Now, I have installed graphene globally. However, when running `graphene` in the command line, I now get:

> /Users/arik/.nvm/versions/node/v4.1.1/bin/graphene: line 1: use strict: command not found
> /Users/arik/.nvm/versions/node/v4.1.1/bin/graphene: line 2: syntax error near unexpected token `('
> /Users/arik/.nvm/versions/node/v4.1.1/bin/graphene: line 2: `var readline = require("readline");'

**Comment by rmhrisk** [comment]
@arik-so it seems maybe we introduced a regression in the graphene CLI, tsc was stripping on only comments but the bits that resulted in node being invoked. We just did a new check in that I hope will resolve this issue.

Can you check:

```
git clone
cd graphene
tsd install && tsc
sudo npm install -g

> graphene
> _
```

**Comment by rmhrisk** [comment]
Reopening, seems there was an issue with the cli.

**Comment by arik-so** [comment]
Hey Ryan,

Please pardon the delay. I ran the steps you you mentioned, but when running `sudo npm install -g`, and I did most definitely not forget the `sudo`, I checked and tried several times, I get the following error:

> \> graphene-pk11@2.0.9 install /Users/arik/.nvm/versions/node/v4.1.1/lib/node_modules/graphene-pk11
> \> tsd install && tsc
> 
> [ERR!] cwd  : /Users/arik/.nvm/versions/node/v4.1.1/lib/node_modules/graphene-pk11
> [ERR!] os   : Darwin 15.5.0
> [ERR!] argv : "/Users/arik/.nvm/versions/node/v4.1.1/bin/node" "/Users/arik/.nvm/versions/node/v4.1.1/bin/tsd" "install"
> [ERR!] node : v4.1.1
> [ERR!] tsd  : 0.6.5
> [ERR!] Error: EACCES: permission denied, open '/Users/arik/.config/configstore/update-notifier-tsd.yml'
> You don't have access to this file.
> 
> [ERR!] CODE : EACCES
> [ERR!] Error: EACCES: permission denied, open '/Users/arik/.config/configstore/update-notifier-tsd.yml'
> You don't have access to this file.
> 
> [ERR!] 
> Please try running this command again as root/Administrator.
> [ERR!] .tsdrc could not be retrieved
> 
> [ERR!] Please include the following file with any support request:
>     /Users/arik/.nvm/versions/node/v4.1.1/lib/node_modules/graphene-pk11/tsd-debug.log
> 
> npm ERR! Darwin 15.5.0
> npm ERR! argv "/Users/arik/.nvm/versions/node/v4.1.1/bin/node" "/Users/arik/.nvm/versions/node/v4.1.1/bin/npm" "install" "-g"
> npm ERR! node v4.1.1
> npm ERR! npm  v2.15.6
> npm ERR! code ELIFECYCLE
> 
> npm ERR! graphene-pk11@2.0.9 install: `tsd install && tsc`
> npm ERR! Exit status 243
> npm ERR! 
> npm ERR! Failed at the graphene-pk11@2.0.9 install script 'tsd install && tsc'.
> npm ERR! This is most likely a problem with the graphene-pk11 package,
> npm ERR! not with npm itself.
> npm ERR! Tell the author that this fails on your system:
> npm ERR!     tsd install && tsc
> npm ERR! You can get information on how to open an issue for this project with:
> npm ERR!     npm bugs graphene-pk11
> npm ERR! Or if that isn't available, you can get their info via:
> npm ERR! 
> npm ERR!     npm owner ls graphene-pk11
> npm ERR! There is likely additional logging output above.
> 
> npm ERR! Please include the following file with any support request:
> npm ERR!     /Users/arik/Developer/graphene/npm-debug.log

It appears to be a permission denied error, but `sudo` is most definitely there.

**Comment by rmhrisk** [comment]
Looks like this is your problem : http://stackoverflow.com/questions/27475480/error-eacces-permission-denied-even-after-using-sudo

Your also running a older version of node, you don't need to update as we test it on Node 4.x but Node 6 is nicer :)

**Comment by microshine** [comment]
You also can remove instal script from package.json, cause you have compiled code

**Comment by rmhrisk** [comment]
You also do not strictly need to have it as a global module

**Comment by arik-so** [comment]
All right,

So I have tried fixing my Node setup by adding the .npmrc file in my ~/ directory, however, it did not yield the promised result of being able to use global installs without root permissions – the error stayed the same.

However, I figured I would also try to run it as root, so I ran `sudo su` and then ran the `npm install -g` script again from the cloned graphene directory, but it yielded the same permission denied error, which appears rather odd considering I was running it as root.

As that didn't work out, I figured I would just install it locally, by running `npm install` in the graphene directory. That worked, but the `graphene` binary is nowhere to be found, though perhaps I'm looking in the wrong places.

I'm sorry all of this is resulting only in one error after the other.

**Comment by rmhrisk** [comment]
This is all so strange :) I will try to reproduce it tonight.

**Comment by arik-so** [comment]
Hey Ryan,

I switched my Node version to 6.x, and reinstalled graphene locally where I was running my script, as opposed to globally. That stopped it from trying to autocomplete the .dylib component, and it was content just reading the .so file.

That almost did the trick. However, running my script with sudo made everything work perfectly. It looks like the .dylib autocompletion is a Node version issue more so than a Graphene issue. I am extremely sorry for all the confusion I have caused.

**Comment by rmhrisk** [comment]
No. We appreciate you going through things as it helps us get our documentation better. I really want to see people use these and the other libraries and good documentation. 

The reason you need to be root is we include SoftHSM for testing purposes which requires root, Graphene does not.

As for your node version issue. In [Travis we pass builds on node 4.4.7](https://travis-ci.org/PeculiarVentures/graphene/jobs/152549457). Maybe the issue you were encountering was fixed between 4.1.1 and 4.4.7, I guess we wont find out right now.

Closing this issue for now, let us know how things go.
